### ISSUE TITLE
High CPU usage when running from Gnome 

### ISSUE THREAD

**Issue by fourier** [exult/exult#315]
Hi, 
Reviving this topic: http://exult.info/forum/viewtopic.php?t=1791251
Running Ubuntu 22.04LTS, default installation: Gnome 42.5, Wayland
Exult:   
```~ /usr/games/exult --version
Exult version 1.6.0
Built at: Apr 26 2020 07:37:09
Compile-time options: USE_TIMIDITY_MIDI, USE_FMOPL_MIDI, USE_ALSA_MIDI, HAVE_ZIP_SUPPORT
Compiler: gcc, version: 9.3.0

Platform: Linux version 5.15.0-58-generic
```
Running from Gnome gives me 100% CPU load, running from Terminal about 25%.

**Comment by alagner** [comment]
What do you mean by CPU usage? CPU usage by exult process, or some other one that spikes the moment exult runs? Can you run top (preferably in the other terminal window, or even better, via ssh) and check what process Is hogging up the CPU? Also, do you use a laptop screen, docking station, what exactly? Dump of pstree would also be nice to see how exactly the processes are run in both cases.

On the forum you've mentioned Debian not behaving that way: did you use Gnome with X or with Wayland on it? TBH, I'd try switching to XServer instead of Wayland if possible and see if that helps.

**Comment by fourier** [comment]
(It was not me in the forum thread, I'm writing about this issue for the first time).
Laptop is macbook air 11" 2013 running on a power cable, no dockstation. Constant CPU usage, not spikes.
![Screenshot from 2023-02-14 08-29-02](https://user-images.githubusercontent.com/900250/218668997-0384d0aa-1431-433a-8485-175183f999d2.png)

[pstree.txt](https://github.com/exult/exult/files/10729851/pstree.txt)

**Comment by i30817** [comment]
Yeah, it's kind of bad. I have a laptop computer from 2018 and i'm at 70-80%. I suspect it's just lack of vsync or something.

**Comment by DominusExult** [comment]
@fourier could you try with current code (also of SDL2)? Version 1.6 is really old by now and SDL2 also went through much more versions.
In current code Vsync is also enabled since 3ff6e0b3d23444443f82d18986e98e23d00afbd4

**Comment by fourier** [comment]
Thanks, I've tried with the latest from master 9ce39c96e963202e6c13cd55378aaf72fb06b21e and same sdl (2.0.20+dfsg-2ubuntu1.22.04.1), its a bit better, now up to 70% of my CPU. 

> @fourier could you try with current code (also of SDL2)? Version 1.6 is really old by now and SDL2 also went through much more versions. In current code Vsync is also enabled since [3ff6e0b](https://github.com/exult/exult/commit/3ff6e0b3d23444443f82d18986e98e23d00afbd4)

**Comment by DominusExult** [comment]
If you turn audio off in exult and set the graphics scaler to point, does it have any effect?

**Comment by fourier** [comment]
No, no effect on neither of these settings.

**Comment by i30817** [comment]
I believe this is mostly a combination of either the BR or HQ scalers and high resolution. You can see the impact of the scalers right away in the menu. Especially if you use a way to limit the cpu, like cpulimit or `systemd-run -p CPUQuota="25%" --scope --uid=$USER --gid=$USER -- exult`

For me, 10% of cpu is the price of mt32 emulation (real) and another 50-60 percent at its worst is the scaler. If you use a cpu limitation thing like the above you can feel the pauses when the buttons appear in the main menu.

**Comment by i30817** [comment]
Could also be xwayland. It may not be relevant, but I build retroarch, and when i was using the x11 version, i noticed that the background animation in there was actually pretty choppy. That went away when i built the wayland version of the thing.

**Comment by i30817** [comment]
I tried it with max resolution for my screen and the point scaler. Interestingly, in game i'm between 15 and 35-40 percent of cpu, depending complexity but the fade in and out appears to jump that to 70, which raises the fan.

The top time function is no longer part of the exult code (it's marked as ??), but from context, it's probably something on sdl or on the gpu driver. Second is `__memcpy_avx_unaligned_erms` and third is `Pentagram::PointScalerInternal<…>::Scale(SDL_Surface*, int, int, int, int, unsigned char*, int, int, int, bool)` again.

Hope this is useful. Probably isn't but it's kind of strange that it's not consistently low even with the point scaler, to the point the fan kicks in.

**Comment by vanfanel** [comment]
CPU usage is VERY high on Debian 12 + Wayland, too. NO XWayland, but Wayland.
Moving the cursor makes it even worse.

I have recorder a video where you can see how bad the problem is.


https://github.com/exult/exult/assets/837585/57709c96-4413-4e03-a6e0-3ef1044286d7



VSYNC is activated.

Seems to be the software scaling. **Please let SDL2 do the scaling, doing software scaling to modern resolutions is an absolute CPU killer**, makes no sense at all.

**Comment by DominusExult** [comment]
How is the CPU usage with SDL3?

**Comment by i30817** [comment]
I'm in android now, it's difficult to tell because my android device at least, doesn't have a fan and apps like htop don't have permission to work (they show nothing of the cpu load).

**Comment by DominusExult** [comment]
> Yeah the SDL based scaler seems to use bilinear filtering . I believe you can tell SDL to use point sampling but we don't at the moment

@Dragon-Baroque if that's the case (which I think I saw something about, too), could both be added as an option? And maybe naming them "SDL Point" & "SDL Bilinear" (instead of taking whatever the backend is named)?

**Comment by wench** [comment]
> > Yeah the SDL based scaler seems to use bilinear filtering . I believe you can tell SDL to use point sampling but we don't at the moment.
> 
> The only attribute that I can find that can apply to `SDL_Surface` scaling and ( in the case of the SDL render scaler ) to the `SDL_Texture` scaling is `SDL_ScaleMode` with values `SDL_SCALEMODE_NEAREST`, `SDL_SCALEMODE_LINEAR` ( the default ), and a new `SDL_SCALEMODE_PIXELART`. Is this what you have in mind ?

Yeah that's what I saw
